sound/soc/sof/ipc4-topology.c

```diff
@@ -1582,55 +1582,92 @@ bool sof_ipc4_copier_is_single_bitdepth(struct snd_sof_dev *sdev,
 	return true;
 }
 
+static int
```

Comment:
@ujfalusi so we're relying on the topology to be always correct here isnt it? What will happen if the topology has only 4ch in all formats but the device only has 2 dmics?

Reply:
Not following the concern @ranj063 ? A bad topology is generally the end of the story, even outside of format changes... We have no error concealment for problems in the topology, and the best solution is to fail big and fail early when a bad topology is encountered.
In this case trying to do a beamforming with only two mics is probably a mistake, isn't it?

Reply:
my concern is related to the number of physical mics on the device. If the topology only has 4ch format but the device only has 2 mics, should we use this topology? If not, how do we know to switch the topology with the 2ch format based on the number of mics on a chromebook. With the HDA topologies, this is easy because we have separate topologies for 2ch and 4ch dmics. But the I2S topologies dont have the distinction right?

Reply:
@ranj063, if we have two dmics and the topology only have 4 channel only then it will not work no matter what you do. You would need 2 channel topology for the two DMICs, no?

I'm not sure if I follow your line of thought.

If we would 'trust' that the blob selection is the correct: the device have 2 DMICs, the loaded topology only have 4 channel configs. We cannot use the loaded topology with the device, right?

Isn't this the same constraint, but in reverse view? If there is no matching blob with the topology then it will not work.

We cannot 'switch' to a new topology, it is expected that the correct topology is loaded for the device, if not, things will not work and they did not worked in the past either?

Reply:
We have the correct topology for microphones count, there can't be a conflict, kernel loads -2ch.tplg or -4ch.tplg based on NHLT information about microphones count. The topologies are generated for 2ch/4ch with help of NUM_DMICS macro, Then in dmic-generic.conf topology I use the same 

```
IncludeByKey.NUM_DMICS {
                                        "2" {}
                                        "4" {}
}
```
to select the dai-copier input format. Each channels count group have the s16/24/32 variants. This is now working well in my tests.

Reply:
To add, the above topology change is only needed when dai-copier's and host-copier's channels counts in connected pipelines can be different. Normally the channels counts are the same.